File: 백준/Gold/1509. 팰린드롬 분할/팰린드롬 분할.py
import sys
input = sys.stdin.readline

word = list(input().rstrip())
n = len(word) # word의 길이 n
pel = [[0] * (n) for _ in range(n)] # 펠린드롬 여부 체크 ex)pel[1][4] 첫번째 부터 4번째까지의 문자가 펠린드롬인지 확인

for i in range(n):
    for j in range(0, n-i):
        start = j
        end = j + i

        if end - start == 0:
            pel[start][end] = 1
            continue
        
        if word[start] == word[end]:
            if end - start == 1 or pel[start+1][end-1] == 1:
                pel[start][end] = 1

dp = [2501] * (n+1)
dp[-1] = 0

for i in range(n):
    for j in range(i+1):
        if pel[j][i] == 1:
            dp[i] = min(dp[i], dp[j-1] + 1)
        else:
            dp[i] = min(dp[i], dp[i-1] + 1)

print(dp[n-1])

File: 백준/Gold/1509. 팰린드롬 분할/팰린드롬 분할.java
import java.util.*;
import java.io.*;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        String s = st.nextToken();
        int length = s.length();
        int[][] pel = new int[length][length];

        for (int i = 0; i < length; i++) {
            for (int j = 0; j < length - i; j++) {
                int start = j;
                int end = j + i;

                if (end - start == 0) {
                    pel[start][end] = 1;
                    continue;
                }

                if (s.charAt(start) == s.charAt(end)) {
                    if (end - start == 1 || pel[start+1][end-1] == 1) {
                        pel[start][end] = 1;
                    }
                }
            }
        }

        int[] dp = new int[length+1];
        Arrays.fill(dp, 2501);
        dp[0] = 0;

        for (int i = 1; i <= length; i++) {
            for (int j = 1; j <= i; j++) {
                if (pel[j-1][i-1] == 1) {
                    dp[i] = Math.min(dp[i], dp[j-1] + 1);
                } else {
                    dp[i] = Math.min(dp[i], dp[i-1] + 1);
                }
            }
        }

        System.out.println(dp[length]);

    }
}
